src/Utils/Rotation.cpp

```diff
@@ -312,15 +312,335 @@ void utils::Rotation::checkUnitary()
 #ifndef SKIP_ASSERT
     double sqrtNorm = static_cast<double>(this->squaredNorm());
     bool checkUnit(fabs(sqrtNorm - 3.) < 1e-4);
-    if (!checkUnit){
-        // Make calulation here
-        utils::Matrix3d mat;
+    if (!checkUnit)
+    {
+
+        // Code addapted from https://github.com/brainexcerpts/3x3_polar_decomposition/blob/master/src/polar_decomposition_3x3.inl
+
+        // Placer inf_norm and one norm in a function
+        // créer fonction polar_decomposition et déplacer dans Matrix3d
+
+        float tolerance = 1e-6;
+        utils::Matrix3d M = this->block(0, 0, 3, 3);
+        utils::Matrix3d R;
+        utils::Matrix3d S;
+        utils::Matrix3d Mk;
+        utils::Matrix3d Ek;
+        float det, E_one_norm;
+        bool use_svd = false;
+
+        
+        // Mk = mat^T
+        Mk = M.transpose();
+        
+        //M one-norm
+        float M_one_norm = 0.0f;
+        for (int i = 0; i < 3; i++)
+        {
+            float col_abs_sum = fabs(M(i, 0)) + fabs(M(i, 1)) + fabs(M(i, 2));
+            if (col_abs_sum > M_one_norm)
+                M_one_norm = col_abs_sum;
+        }
+
+        //M infinity-norm
+        float M_inf_norm = 0.0;
+        for (int i = 0; i < 3; i++)
+        {
+            float row_sum = fabs(M(i, 0)) + fabs(M(i, 1)) + fabs(M(i, 2));
+            if (row_sum > M_inf_norm)
+                M_inf_norm = row_sum;
+        }
+
+        do
+        {
+            utils::Matrix3d M_adj_Tk;
+
+            /*
+            // row 2 x row 3
+            cross(&(Mk[3]), &(Mk[6]), &M_adj_Tk(0, 0)));
+            // row 3 x row 1
+            cross_product(&(Mk[6]), &(Mk[0]), &(M_adj_Tk[3]));
+            // row 1 x row 2
+            cross_product(&(Mk[0]), &(Mk[3]), &(M_adj_Tk[6]));
+            */
+
+            det = Mk(0, 0) * M_adj_Tk(0, 0) + Mk(0, 1) * M_adj_Tk(0, 1) + Mk(0, 2) * M_adj_Tk(0, 2);
+
+            if( det <= 1e-6 )
+            {
+                use_svd = true;
+                std::cout << "Oups, SVD serait utile!\n";
+                break;
```

Comment:
Is there a reason not to do svd in the first place?

I am not sure it makes sense to write anything to the console here... But if it make sense, please use:
`utils::Error::warning(false, "message");`

Reply:
It makes sense for debugging ;p